### ISSUE TITLE
Sample never completes if the sampler sequence completes before the sampled sequence

### ISSUE THREAD

**Issue by glopesdev** [Reactive-Extensions/Rx.NET#179]
Is this expected behavior? It looks to me that if sampler terminates, then there is a guarantee that there will be no more values coming out of **Sample**, and that it could be safely terminated. I was surprised to find that it never terminates. Is there some problematic scenario where the current behavior would be useful?

To reproduce:

```c#
var data = Observable.Timer(TimeSpan.Zero, TimeSpan.FromSeconds(0.5));
var sampler = Observable.Timer(TimeSpan.FromSeconds(1));
var result = data.Sample(sampler).Materialize();
result.Subscribe(Console.WriteLine);
Console.ReadLine();
```

**Comment by shiftkey** [comment]
I think this is due to the `data` observable never completing, and is by design. Here's a snapshot of how it looks as an Rx marble diagram:

<img width="893" src="https://cloud.githubusercontent.com/assets/359239/15095934/8b02c4e0-1525-11e6-8f39-481ed34b8080.png">

Note where the sequence terminates...

**Comment by glopesdev** [comment]
Actually this happens independently of whether or not the sampled sequence terminates (I've updated my example with **Take**).

If this is by design, then my next question is: why? Is there any situation in which it would make sense to have this "by design"? I don't like the idea of operators that promote dangling sequences, and in general Rx seems to be quite conservative in this respect. Most operators terminate as soon as they can reasonably be allowed to terminate in order to make sure resource allocation side-effects are disposed of.

That's very reasonable and also particularly important when mixing observable streams with operators that convert observables from infinite/finite and hot/cold such as Window or Replay. I don't want racing conditions changing the behavior of my slices where some processing terminates and others don't because the last sampling tick came at the wrong time...

**Comment by shiftkey** [comment]
Just pointed two observers at each sequence (this is on Rx 2.2.5 built from source) and the sampler's subscription doesn't raise `OnCompleted`. Interesting.

<img width="186" src="https://cloud.githubusercontent.com/assets/359239/15097365/40e0e64e-155c-11e6-8a38-067bc7509697.png">

**Comment by glopesdev** [comment]
You mean the result from **Sample** doesn't raise `OnCompleted`, right? The sampler sequence itself should complete.

Here's my own print out of all three sequences, which agrees with yours if you replace **Sampler** by **Result**.

![sample](https://cloud.githubusercontent.com/assets/5315880/15097497/233be77a-1515-11e6-9271-26b63eafb80e.png)

**Comment by shiftkey** [comment]
@glopesdev I edited my previous comment to show the full sample (including subscriptions)

**Comment by glopesdev** [comment]
@shiftkey great, we're in agreement then, although I would rename the second `WriteLine` to print **"Sample"** instead of **"Sampler"**, since it is effectively printing out the result from the **Sample** operator.

I have just run the sample against the latest head and the basic point of the issue remains: **Sample** never completes if the sampler sequence terminates first.

**Comment by RxDave** [comment]
Wow I wouldn't have expected that. Definitely a bug.

**Comment by RxDave** [comment]
To be clear, I still think that OnCompleted shouldn't occur exactly when the sampler calls OnCompleted, but when all sequences have OnCompleted.

OnCompleted carries information. In this case, it should, assuming the bug has been fixed, tell observers when the source and the sampler have both OnCompleted. The reason is simply that this overload samples the notifications of a given observable based on the timing of another. It would actually seem strange to me for the sampler to say when to sample the latest OnNext from the source, but when OnCompleted arrives, instead of sampling OnCompleted from the source (which basically just means to wait for the source to complete naturally), but what you're assuming is that it will simply terminate the output immediately, ignoring the source's actual OnCompleted. I can understand your point of view, that it seems wasteful to have a bunch of silence between the sampler and the OnCompleted, but I think that's just an inconsistent way of looking at this operator's semantics. I.e., I see it as the sampler says when to sample the source, if it goes away then you just won't receive any more sampled notifications, you'll just *may* eventually get the source's OnCompleted.

**Comment by RxDave** [comment]
```c#
sampler.Publish(s => xs.Sample(s).TakeUntil(s));
```

**Comment by glopesdev** [comment]
@RxDave thanks for the clarification, I would be happy with the semantics you described.

Can you give some context to the code sample you posted? Initially I thought it would be some kind of workaround for having **Sample** terminate when the sampler completes, but then I guess you would need to use `TakeUntil(s.TakeLast(1))`, so maybe you had something else in mind?

**Comment by RxDave** [comment]
My mistake, thanks for the correction. I've added `IgnoreElements` to fix the problem.

**Comment by glopesdev** [comment]
@RxDave  I don't think `IgnoreElements` is enough because `TakeUntil` ignores completion messages for the purposes of terminating the sequence (i.e. only **OnNext** calls are considered). I've had this problem before and I converged on using `TakeLast(1)` because at least it ensures that the termination token is sent out only when the sequence completes. It doesn't work for empty sequences though, which is annoying...

**Comment by RxDave** [comment]
Hmm, you're right. I've thought for the longest time that `TakeUntil` reacts to `OnCompleted` as well. I wonder if they changed it at some point and I just never noticed. That's good to know.

**Comment by RxDave** [comment]
Here's an alternative:
`sampler.Publish(s => xs.Sample(s).TakeUntil(s.DefaultIfEmpty()));`

**Comment by RxDave** [comment]
Yeah, I think I'll use DefaultIfEmpty from now on. It has the nice additional effect of ensuring that OnCompleted -> OnNext, if for any reason the source is an empty sequence.

**Comment by RxDave** [comment]
Oops, forgot `TakeLast(1)` again. Corrected.

**Comment by glopesdev** [comment]
@RxDave cool, I never noticed `DefaultIfEmpty`. However, it looks like you still need to couple it with a `TakeLast(1)`, because if the sequence is non-empty it just starts streaming all the elements from the beginning, which is not what you want if you want to "TakeUntil the end".

However, after checking just now I think I found an even better alternative: `LastOrDefaultAsync` returns the last element of an observable sequence or a default value if no such element exists. I think I'll use this one from now on, nice!

**Comment by RxDave** [comment]
Oooh, `LastOrDefaultAsync`, that's it.

**Comment by shiftkey** [comment]
@glopesdev @RxDave is this something something we should actually fix?

**Comment by RxDave** [comment]
Hmm, it's a breaking change and there's a reasonable workaround. Perhaps it's not worth fixing at this point?

**Comment by glopesdev** [comment]
@shiftkey Yes, as it was clearly identified as being a bug. I recommend adopting the behavior suggested by @RxDave above. I can make a pull request with unit tests, but maybe it's best for @bartdesmet or @mattpodwysocki to weigh in?

**Comment by glopesdev** [comment]
@RxDave You're right about that, but the new major release 3.0.0 would be a good opportunity to fix this, especially as Rx.NET is expanding it's reach to even more platforms. You'll just be propagating the bug further and further into the future, which will only make the problem worse...

**Comment by RxDave** [comment]
@glopesdev Makes sense to me.

I think what we really need is an additional sampling/throttling library that rounds out the missing behaviors in a set of well-designed, generalized operators. I mean, how many people have written their own "responsive" sampling operator? I.e., Take the first, then sample when there's chattiness and reset when there's silence. There are like dozens out there.

**Comment by glopesdev** [comment]
@RxDave guilty as charged...
`source.Window(sampler).SelectMany(xs => xs.Take(1))`

**Comment by danielcweber** [comment]
@glopesdev @RxDave @ghuntley 
Should we fix it now and go for the breaking change? IMO the described behaviour is clearly unexpected, so the change might not be that breaking after all.

**Comment by glopesdev** [comment]
@danielcweber 4.0 is technically a major release, so we should definitely use it for behaviour changes of this kind. In fact, this is actually just a bug fix according to semantic versioning, since no APIs are actually breaking. As you say, there was a failure to comply with the documented behaviour of the operator, as also acknowledged above.

**Comment by onovotny** [comment]
I think we can take this fix for 4.0. Is there a PR for this against master?

**Comment by glopesdev** [comment]
@danielcweber sounds good, I don't mind contributing some tests if necessary

**Comment by danielcweber** [comment]
#274 deals with periodic sampling by a timer. There may be a bug there as well but this issue deals with the scenario where you have a sampling sequence. @glopesdev A failing test for the desired scenario would be awesome.

**Comment by danielcweber** [comment]
@onovotny @glopesdev  https://github.com/dotnet/reactive/pull/520
